functions/export-csv.js

```diff
@@ -15,79 +15,84 @@
  * limitations under the License.
  */
 
-'use strict';
+"use strict";
 
-const {db} = require('./common/context');
+const csv = require("@fast-csv/format");
+const { db } = require("./common/context");
 
-function exportCsv(req, res) {
-  const {
-    project: projectId,
-    featureType: featureTypeId,
-    lang: desiredLanguage,
-  } = req.query;
-  // TODO: Simplify/refactor post G4G19.
-  var elements = [];
-  var features = [];
-  return db.fetchProject(projectId).then(
-    project => {
-      if (!project.exists) {
-        res.status(404).send('not found');
-        return Promise.reject(new Error('project not found: ' + projectId))
-      }
-      res.type('text/csv');
-      let form = project.get('featureTypes')[featureTypeId]['forms'];
-      elements = form[Object.keys(form)[0]]['elements'];
-      var arr = []
-      for (var el in elements) {
-        arr.push('"' + escapeQuotes(elements[el]['labels']['_']) + '"');
-      }
-      arr.push('"Latitude"');
-      arr.push('"Longitude"');
-      res.write(arr.join(',') + '\n');
-      return db.fetchFeatures(project.id);
-    }
-  ).then(
-    data => {
-      features = data;
-      var promises = [];
-      features.docs.forEach(feature => {
-        promises.push(db.fetchRecords(projectId, feature.id));
-      });
-      return Promise.all(promises);
-    }
-  )
-  .then(
-    records => {
-      var records = records[0]; // TODO: remove this line if fetchRecords give correct data
-      features.docs.forEach(feature => {
-          records.docs.forEach(record => {
-            if (record.get('featureId') == feature.id) { // TODO: remove this condition if fetchRecords give correct data
-              var arr = [];
-              for (var el in elements) {
-                var value = record.data()['responses'][elements[el]['id']];
-                value = value == (undefined || null) ? '' : value;
-                arr.push('"' + escapeQuotes(value) + '"');
-              }
-              var center = feature.data()['center'];
-              arr.push('"' + center['_latitude'] + '"');
-              arr.push('"' + center['_longitude'] + '"');
-              res.write(arr.join(',') + '\n');
-            }
-          });
-      });
-      return res.end();
-    }
-  )
-  .catch(
-    err => {
-      console.error("Export Failed: " , err);
-      return res.status(500).end();
-    }
-  )
-
-  function escapeQuotes(str) {
-    return str.replace(/\r?\n/g, '\\n').replace(/"/g, '""');
+async function exportCsv(req, res) {
+  const { project: projectId, layer: layerId } = req.query;
+  const project = await db.fetchProject(projectId);
```

Comment:
We're only loading one project here, but project contains multiple layers - were you referring to those? We could fetch only one layer, but they're very small and it would make the implementation more complicated. In future the project JSON might also contain settings or metadata that would be needed for the export.

Reply:
Yes sorry was referring to layers here. Why would fetching the relevant layer make the implementation more complicated? You can get rid of a few lines here. no?

Reply:
I'm not really sure how to read a partial document in Firestore - after a quick scan of https://firebase.google.com/docs/firestore/query-data/get-data and related articles I didn't find it.

This is a temporary sol'n for our demo next week. In the production version, this will probably be replaced with an aptly scoped gRPC call (getLayerConfig?). Can we leave this as is for now and do it "the right way" when we get to a production version?

Reply:
Ok let's unblock this for now.

Reply:
Thanks!